onnxruntime/contrib_ops/cpu/math/sparse_add.cc

```diff
@@ -0,0 +1,227 @@
+      auto B_ind_span = GetIndicesSpan(input_B, B_converted_indices);
+
+      enum Source {
+        kInputA = 1,
+        kInputB = 2,
+        kInputSum = 3
+      };
+
+      std::vector<std::tuple<int64_t, Source>> result_indices;
+      result_indices.reserve(A_ind_span.size() + B_ind_span.size());
+      size_t a_ind = 0;
+      size_t b_ind = 0;
+
+      const auto a_span_size = A_ind_span.size();
+      const auto b_span_size = B_ind_span.size();
+      while (a_ind < a_span_size || b_ind < b_span_size) {
+        if (a_ind < a_span_size && b_ind < b_span_size) {
+          auto a_val = A_ind_span[a_ind];
+          auto b_val = B_ind_span[b_ind];
+          if (a_val == b_val) {
+            result_indices.push_back(std::make_tuple(a_val, kInputSum));
+            a_ind++;
+            b_ind++;
+          } else if (a_val < b_val) {
+            result_indices.push_back(std::make_tuple(A_ind_span[a_ind++], kInputA));
+          } else {
+            result_indices.push_back(std::make_tuple(B_ind_span[b_ind++], kInputB));
+          }
+        } else if (a_ind < a_span_size) {
+          result_indices.push_back(std::make_tuple(A_ind_span[a_ind++], kInputA));
+        } else {
+          result_indices.push_back(std::make_tuple(B_ind_span[b_ind++], kInputB));
+        }
+      }
+
+      const auto element_size = input_A.DataType()->AsPrimitiveDataType()->Size();
+      sparse_utils::CopyElementFunc copy_func;
+      switch (element_size) {
+        case 1: {
+          copy_func = sparse_utils::CopyElementAligned<uint8_t>;
+          break;
+        }
+        case 2: {
+          copy_func = sparse_utils::CopyElementAligned<uint16_t>;
+          break;
+        }
+        case 4: {
+          copy_func = sparse_utils::CopyElementAligned<uint32_t>;
+          break;
+        }
+        case 8: {
+          copy_func = sparse_utils::CopyElementAligned<uint64_t>;
+          break;
+        }
+        default:
+          return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                                 "Element_size of: ", element_size,
+                                 " is not supported.", " data_type: ", input_A.GetElementType());
+      }
+
+      const auto* values_A = input_A.Values().DataRaw();
+      size_t a_src_idx = 0;
+      const auto* values_B = input_B.Values().DataRaw();
+      size_t b_src_idx = 0;
+      auto coo_mutator = output.MakeCooData(result_indices.size(), result_indices.size());
+      auto* output_values = coo_mutator.Values().MutableDataRaw();
+      size_t dst_idx = 0;
+      auto* output_indices = coo_mutator.Indices().MutableData<int64_t>();
+      utils::MLTypeCallDispatcherFromTypeList<element_type_lists::AllFixedSizeExceptHalf>
+          t_disp(input_A.GetElementType());
+      for (const auto& e : result_indices) {
+        auto where = std::get<1>(e);
+        switch (where) {
+          case kInputA:
+            copy_func(output_values, values_A, dst_idx, a_src_idx);
```

Comment:
Have a naive question. Why not access output_values and values_A directly such as `*output_values++ = *values_A++` , or `*output_values++ = *values_A++ + *values_B++` , etc. 
Fromt he copy_func() it seems we need some alignment, but from the Sum class it seems we can use array access directly so I was wondering whether we can do direct access or not...

Reply:
We can do direct access at the cost of type depended templated code with increased binary size.